test/test_closure_pte.cpp

```diff
@@ -0,0 +1,143 @@
+    constexpr Real b = 0.58;
+    constexpr Real k = 1.35;
+    constexpr Real n = 2.66182;
+    constexpr Real vc = 0.75419;              // cm^3 / g
+    constexpr Real pc = 3.2 * GPa;            // microbar
+    constexpr Real Cv = 0.001072 * MJ_per_kg; // erg / g / K
+    constexpr Real Q = 4.115 * MJ_per_kg;
+    eos_arr[2] = DavisProducts(a, b, k, n, vc, pc, Cv).Modify<ShiftedEOS>(-Q);
+
+    // Move EOS array from host to device
+    for (auto eos : eos_arr) {
+      eos.GetOnDevice();
+    }
+    constexpr size_t EOS_bytes = num_eos * sizeof(EOS);
+    EOS *v_EOS = (EOS *)PORTABLE_MALLOC(EOS_bytes);
+    portableCopyToDevice(v_EOS, eos_arr.data(), EOS_bytes);
+
+    GIVEN("A difficult state for the density-temperature PTE solver") {
+      // Define the state
+      constexpr Real spvol_bulk = 6.256037280402093e-01;
+      constexpr Real sie_bulk = -1.441692060406520e+10;
+      const std::array<const Real, num_eos> mass_frac = {
+          1.10382442033331e-10, 0.124935312146569, 0.875064687743048};
+
+      // Initial guess: all materials at the cell density
+      const std::array<const Real, num_eos> vol_frac = mass_frac;
+
+      // Calculate material densities (and corresponding energies) and the total
+      // volume fraction
+      std::array<Real, num_eos> densities;
+      std::array<Real, num_eos> sies;
+      Real vfrac_sum = 0;
+      for (auto i = 0; i < num_eos; ++i) {
+        densities[i] = mass_frac[i] / spvol_bulk / vol_frac[i];
+        sies[i] = sie_bulk * mass_frac[i];
+        vfrac_sum += vol_frac[i];
+      }
+
+      // Initialize pressure and temperature arrays to zero
+      std::array<Real, num_eos> temperatures = {0., 0., 0.};
+      std::array<Real, num_eos> pressures = {0., 0., 0.};
+
+      // Copy values to device (when available)
+      constexpr size_t bytes = num_eos * sizeof(Real);
+      Real *v_densities = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_densities, densities.data(), bytes);
+      Real *v_vol_frac = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_vol_frac, vol_frac.data(), bytes);
+      Real *v_sies = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_sies, sies.data(), bytes);
+      Real *v_temperatures = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_temperatures, temperatures.data(), bytes);
+      Real *v_pressures = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_pressures, pressures.data(), bytes);
+
+      THEN("The PTE solver should converge") {
+        // Allocate scratch space for the PTE solver
+        const int pte_solver_scratch_size = PTESolverRhoTRequiredScratch(num_eos);
+        const size_t scratch_bytes = pte_solver_scratch_size * sizeof(Real);
+        double *scratch = (double *)PORTABLE_MALLOC(scratch_bytes);
+        constexpr size_t lambda_bytes = MAX_NUM_LAMBDAS * num_eos * sizeof(Real);
+        double **lambdas = (double **)PORTABLE_MALLOC(lambda_bytes);
```

Comment:
doesn't quite look right here. This is roughly creating an array of pointers of length MAX_NUM_LAMBDAS * num_eos, though not quite b/c the underlying type should be Real*. Creating the equivalent of a matrix with **'s is a somewhat complex procedure. First one mallocs an array of pointers (the ** quantity), then for each pointer, one does a malloc of the actual underlying type. And the frees need to handled in the same way.

Reply:
Ah fair enough. I was trying to do it in one quick allocation, but I also need to populate the pointers. I'll change this.

Reply:
Let me know if I'm doing this right now